presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+                        schemaTableName,
+                        recursiveDeleteResult.getNotDeletedEligibleItems()));
+            }
+        });
+    }
+
+    public synchronized Optional<List<String>> getPartitionNames(String databaseName, String tableName)
+    {
+        return doGetPartitionNames(databaseName, tableName, Optional.empty());
+    }
+
+    public synchronized Optional<List<String>> getPartitionNamesByParts(String databaseName, String tableName, List<String> parts)
+    {
+        return doGetPartitionNames(databaseName, tableName, Optional.of(parts));
+    }
+
+    private Optional<List<String>> doGetPartitionNames(String databaseName, String tableName, Optional<List<String>> parts)
+    {
+        checkReadable();
+        Optional<Table> table = getTable(databaseName, tableName);
+        if (!table.isPresent()) {
+            return Optional.empty();
+        }
+        List<String> partitionNames;
+        TableSource tableSource = getTableSource(databaseName, tableName);
+        switch (tableSource) {
+            case CREATED_IN_THIS_TRANSACTION:
+                partitionNames = ImmutableList.of();
+                break;
+            case PRE_EXISTING_TABLE: {
+                Optional<List<String>> partitionNameResult;
+                if (parts.isPresent()) {
+                    partitionNameResult = delegate.getPartitionNamesByParts(databaseName, tableName, parts.get());
+                }
+                else {
+                    partitionNameResult = delegate.getPartitionNames(databaseName, tableName);
+                }
+                if (partitionNameResult.isPresent()) {
+                    partitionNames = partitionNameResult.get();
+                }
+                else {
+                    throw new PrestoException(TRANSACTION_CONFLICT, "Table %s.%s was dropped by a different transaction.");
+                }
+                break;
+            }
+            default:
+                throw new UnsupportedOperationException("Unknown table source");
+        }
+        Map<List<String>, Action<PartitionAndMore>> partitionActionsOfTable = partitionActions.computeIfAbsent(new SchemaTableName(databaseName, tableName), k -> new HashMap<>());
+        ImmutableList.Builder<String> resultBuilder = ImmutableList.builder();
+        // alter/remove newly-altered/dropped partitions from the results from underlying metastore
+        for (String partitionName : partitionNames) {
+            List<String> partitionValues = toPartitionValues(partitionName);
+            Action<PartitionAndMore> partitionAction = partitionActionsOfTable.get(partitionValues);
+            if (partitionAction == null) {
+                resultBuilder.add(partitionName);
+                continue;
+            }
+            switch (partitionAction.getType()) {
+                case ADD:
+                    throw new PartitionAlreadyExistsException(new SchemaTableName(databaseName, tableName), Optional.of(partitionValues));
```

Comment:
Shouldn't this be a transaction conflict... another transaction created the partition